desired_state_generator/src/gen_desired_state.py

```diff
@@ -0,0 +1,205 @@
+# License for the specific language governing permissions and limitations
+# under the License.
+#
+# SPDX-License-Identifier: Apache-2.0
+
+import argparse
+import hashlib
+import json
+import os
+import re
+from typing import Any, Dict, List, Optional
+
+import velocitas_lib
+import velocitas_lib.services
+
+VSS_SOURCE_DEFAULT = "vss-source-default-vss"
+VSS_SOURCE_CUSTOM = "vss-source-custom-vss"
+DATABROKER_ID = "data-broker-grpc"
+GRPC_INTERFACE_ID = "dataprovider-proto-grpc"
+DATABROKER_PACKAGE_ID = "vehicledatabroker"
+MQTT_PACKAGE_ID = "mqtt-broker"
+
+VSS_INTERFACE = "vehicle-signal-interface"
+PUBSUB_INTERFACE = "pubsub"
+GRPC_INTERFACE = "grpc-interface"
+
+
+def is_uri(path: str) -> bool:
+    """Check if the provided path is a URI.
+
+    Args:
+        path (str): The path to check.
+
+    Returns:
+        bool: True if the path is a URI. False otherwise.
+    """
+    return re.match(r"(\w+)\:\/\/(\w+)", path) is not None
+
+
+def parse_vehicle_signal_interface(config: Dict[str, Any]) -> List[str]:
+    """Parse the vehicle signal interface config.
+
+    Args:
+        config (Dict[str, Any]): The json-config of the interface,
+        as defined in the appManifest.json.
+
+    Returns:
+        List[str]: A list of requirements defined by the config.
+    """
+    requirements = []
+    src = str(config["src"])
+    vss_release_prefix = (
+        "https://github.com/COVESA/vehicle_signal_specification/releases/download/"
+    )
+    version = ""
+    if vss_release_prefix in src:
+        version = src.removeprefix(vss_release_prefix).split("/")[0]
+        requirements.append(f"{VSS_SOURCE_DEFAULT}:{version}")
+    else:
+        version = get_md5_from_file_content(src)
+        requirements.append(f"{VSS_SOURCE_CUSTOM}:{version}")
+
+    requirements.append(f"{DATABROKER_ID}:{get_package_version(DATABROKER_PACKAGE_ID)}")
+
+    datapoints = config["datapoints"]["required"]
+    for datapoint in datapoints:
+        path = str(datapoint["path"]).lower().replace(".", "-")
+        access = datapoint["access"]
+        requirements.append(f"vss-{access}-{path}:{version}")
+
+    return requirements
+
+
+def parse_grpc_interface(config: Dict[str, Any]) -> str:
+    """Parse the grpc interface config.
+
+    Args:
+        config (Dict[str, Any]): The json-config of the interface,
+        as defined in the appManifest.json.
+
+    Returns:
+        str: The requirement with md5-hash of the proto-file as version.
+    """
+    src = str(config["src"])
+
+    return f"{GRPC_INTERFACE_ID}:{get_md5_from_file_content(src)}"
+
+
+def get_md5_from_file_content(src: str) -> str:
+    """Get the md5-hash of the contents of a file defined by a source.
+
+    Args:
+        src (str): The source of the file. Can either be a local file-path or an URI
+
+    Returns:
+        str: The md5-hash of the file.
+    """
+    file_path = src
+    if is_uri(src):
+        file_path = "./tmp"
```

Comment:
better use `tempfile.TemporaryDirectory` which will auto remove the directory and its contents once the object is cleaned up.